zipline/assets/futures.py

```diff
@@ -0,0 +1,165 @@
+from pandas import Timestamp, Timedelta
+from pandas.tseries.tools import normalize_date
+
+
+class FutureChain(object):
+    """ Allows users to look up future contracts.
+
+    Parameters
+    ----------
+    asset_finder : AssetFinder
+        An AssetFinder for future contract lookups, in particular the
+        AssetFinder of the TradingAlgorithm instance.
+    get_datetime : function
+        A function that returns the simulation datetime, in particular
+        the get_datetime method of the TradingAlgorithm instance.
+    root_symbol : str
+        The root symbol of a future chain.
+    as_of_date : pandas.Timestamp, optional
+        Date at which the chain determination is rooted. I.e. the
+        existing contract that expires first after (or on) this date is
+        the primary contract, etc. If not provided, the current
+        simulation date is used as the as_of_date.
+
+    Attributes
+    ----------
+    root_symbol : str
+        The root symbol of the future chain.
+    as_of_date
+
+    Methods
+    -------
+    as_of(dt)
+    offset(time_delta)
+
+    Raises
+    ------
+    RootSymbolNotFound
+        If the FutureChain is initialized with a root symbol for which
+        a future chain could not be found.
+    """
+    def __init__(self, asset_finder, get_datetime, root_symbol,
+                 as_of_date=None):
+        self.root_symbol = root_symbol
+
+        # Reference to the algo's AssetFinder for contract lookups
+        self._asset_finder = asset_finder
+        # Reference to the algo's get_datetime to know the current dt
+        self._algorithm_get_datetime = get_datetime
+
+        # If an as_of_date is provided, self._as_of_date uses that
+        # value, otherwise None. This attribute backs the as_of_date property.
+        if as_of_date:
+            self._as_of_date = normalize_date(Timestamp(as_of_date, tz='UTC'))
+        else:
+            self._as_of_date = None
+
+        # Attribute to cache the most up-to-date chain, and the dt when it was
+        # last updated.
+        self._current_chain = []
+        self._last_updated = None
+
+        # Get the initial chain, since self._last_updated is None.
+        self._maybe_update_current_chain()
+
+    def __repr__(self):
+        if self._as_of_date:
+            return "FutureChain(root_symbol='%s', as_of_date='%s')" % (
+                self.root_symbol, self.as_of_date)
+        else:
+            return "FutureChain(root_symbol='%s')" % self.root_symbol
+
+    def _get_datetime(self):
+        """
+        Returns the normalized simulation datetime.
+
+        Returns
+        -------
+        pandas.Timestamp
+            The normalized datetime of FutureChain's TradingAlgorithm.
+        """
+        return normalize_date(
+            Timestamp(self._algorithm_get_datetime(), tz='UTC')
+        )
+
+    @property
+    def as_of_date(self):
```

Comment:
@jfkirk the example @yankees714 posted is what I had in mind.  Being able to say "interpret FutureChain X in another `as_of`" seems like a useful idiom.